### DISCUSSION TITLE
put a ruleset file into a new directory such as `/data/config/init.json`

### DISCUSSION THREAD

**Discussion by MonicaisHer** [lf-edge/ekuiper#1600] [Q&A] [answered]
Hello, is that possible to put a ruleset file into a new directory instead of the default [`/data`](https://github.com/lf-edge/ekuiper/blob/master/docs/en_US/operation/config/configuration_file.md#ruleset-provision) directory? For example, `/data/config/init.json`. I have read ekuiper's documentation and couldn't find a related solution. Thank you in advance.

**Comment by ngjaying** [comment, marked as answer]
Hello Monica,
This is not configurable now. Any reason to change the location?

**Reply by MonicaisHer** [reply]
Hi ngjaying, thanks for your reply. I am using a [content interface](https://snapcraft.io/docs/content-interface) to deliver rulesets file `init.json` from a provider to a consumer (ekuiper). 

The type path of the provider and consumer could only be directory, so here is my original approach:
provider path: `/config` 
consumer path: `/ekuiper/data`
rulesets path after connecting provider and consumer: `/ekuiper/data/config/init.json`

This was the reason I was wondering if I could configure the default location to `/ekuiper/data/config`.

Fortunately, there is a solution for my issue without configuring it:
provider path: `/data`
consumer path: `/ekuiper`
rulesets path: `/ekuiper/data/init.json`

**Reply by farshidtz** [reply]
I think the main issue relates to eKuiper using the `${KuiperBaseKey}/data` directory to both read the `init.json` and add system generated files.

If the directory is provided from outside (e.g. a read-only Docker volume), then eKuiper will fail to deploy system-generated files alongside `init.json`. Ideally, eKuiper should make the path to ruleset init file configuration so that it can be sourced from a readonly location along with other user-defined configuration files.

**Reply by ngjaying** [reply]
The reason to put it in `data` as the convention is that we suppose this `init.json` to be a one time configuration and only apply to fresh eKuiper instance. Given a fresh instance, there should be nothing inside `data` so that users can mount it to an empty directory with an optional `init.json` file. After running, the `data` directory will contain all persisted user data. In order to persist the user data, the directory are usually mounted to a writable external folder which is the default behavior in edgex docker compose. I think it is fine to have `init.json` together with other system generated user data along running. The problem is just whether you can use a writable instead of readonly Docker volume to mount `data`.
